app/controllers/api/v0/api_controller.rb

```diff
@@ -40,37 +40,39 @@ def authenticate!
       @user = User.find(doorkeeper_token.resource_owner_id)
       return error_unauthorized unless @user
     elsif request.headers["api-key"]
-      authenticate_with_api_key!
+      @user = authenticate_with_api_key
+      return error_unauthorized unless @user
     elsif current_user
       @user = current_user
     else
       error_unauthorized
     end
   end
 
+  # Checks if the user is authenticated, sets @user to nil otherwise
+  def authenticate_with_api_key_or_current_user
```

Comment:
There's now two `#authenticate_with_api_key_or_current_user`?  see line:59

Reply:
one is with the bang, the other is without it :D

Reply:
Oh, this is not the first time my eye misses the bang 😅

Reply:
@rhymes In that case we maybe should change the second method, to avoid any drift, as unlikely as that may seem:

```ruby
  def authenticate_with_api_key_or_current_user!
    authenticate_with_api_key_or_current_user
    error_unauthorized unless @user
  end
```

Reply:
I kept it explicit mostly because it wasn't enough duplication to reuse :D

Reply:
I'm less concerned about duplication and more about someone changing one place but not the other. Also, most of the Rails defined bang-methods call the "normal" one and then raise, e.g. 

```ruby
def take!
  take || raise_record_not_found_exception!
end
```

Either way, it's no big deal and I'll leave it to your judgement.